public/app/features/explore/utils/supplementaryQueries.ts

```diff
@@ -129,98 +127,66 @@ const getSupplementaryQueryFallback = (
-      const targets = sets[key];
-      mixed.push({
-        datasource: getDataSourceSrv().get(targets[0].datasource, request.scopedVars),
-        targets,
-      });
+  const providers = datasources.map(({ datasource, targets }, i) => {
+    const dsRequest = cloneDeep(request);
+    dsRequest.requestId = `${dsRequest.requestId || ''}_${i}`;
+    dsRequest.targets = targets;
+
+    if (hasSupplementaryQuerySupport(datasource, type)) {
+      return datasource.getDataProvider(type, dsRequest);
+    } else {
+      return getSupplementaryQueryFallback(type, explorePanelData, targets, datasource.name);
     }
+  });
 
-    return from(mixed).pipe(
-      mergeMap((query, i) => {
-        return from(query.datasource).pipe(
-          mergeMap((ds) => {
-            const dsRequest = cloneDeep(request);
-            dsRequest.requestId = `mixed-${type}-${i}-${dsRequest.requestId || ''}`;
-            dsRequest.targets = query.targets;
-
-            if (hasSupplementaryQuerySupport(ds, type)) {
-              const dsProvider = ds.getDataProvider(type, dsRequest);
-              if (dsProvider) {
-                // 1) It provides data for current request - use the provider
-                return dsProvider;
-              } else {
-                // 2) It doesn't provide data for current request -> return nothing
-                return of({
-                  data: [],
-                  state: LoadingState.NotStarted,
-                });
-              }
-            } else {
-              // 3) Data source doesn't support the supplementary query -> use fallback
-              // the fallback cannot determine data availability based on request, it
-              // works on the results once they are available so it never uses the cache
-              return getSupplementaryQueryFallback(type, explorePanelData, query.targets, ds.name);
-            }
-          })
-        );
-      }),
-      scan<DataQueryResponse, DataQueryResponse>(
-        (acc, next) => {
-          if (acc.error || next.state === LoadingState.NotStarted) {
-            return acc;
-          }
+  const definedProviders: Array<Observable<DataQueryResponse>> = providers.filter(
+    (item: Observable<DataQueryResponse> | undefined): item is Observable<DataQueryResponse> => {
+      return !!item;
+    }
+  );
 
-          if (next.state === LoadingState.Loading && acc.state === LoadingState.NotStarted) {
-            return {
-              ...acc,
-              state: LoadingState.Loading,
-            };
-          }
+  if (definedProviders.length === 0) {
+    return undefined;
+  } else if (definedProviders.length === 1) {
+    return definedProviders[0];
+  }
 
-          if (next.state && next.state !== LoadingState.Done) {
-            return acc;
-          }
+  return merge(...definedProviders).pipe(
+    scan<DataQueryResponse, DataQueryResponse>(
+      (acc, next) => {
+        if (acc.error || next.state === LoadingState.NotStarted) {
```

Comment:
`error` is deprecated, should we also check for `errors`?

Reply:
Switched to `errors` in 2f5a09453a865a359266b3ed9bfea8eca7f43c48.